### PR TITLE
21.1 Known Limitations

### LINKED ISSUE #10346: Document 21.1 known limitations
https://github.com/cockroachdb/cockroach/issues?q=is%3Aopen+label%3Adocs-known-limitation+-label%3Adocs-done

Also look at https://github.com/cockroachdb/docs/issues?q=is%3Aopen+is%3Aissue+label%3AT-known-limitation.

And update any known limitations that have been resolved (remove or mark as resolved).

### PR DESCRIPTION
**Pull request by ericharmeling**
Fixes https://github.com/cockroachdb/docs/issues/9396.
Fixes https://github.com/cockroachdb/docs/issues/7524.
Fixes https://github.com/cockroachdb/docs/issues/7188.
Fixes https://github.com/cockroachdb/docs/issues/8955.
Fixes #10346.

Related to https://github.com/cockroachdb/cockroach/issues/59649.
Related to https://github.com/cockroachdb/cockroach/issues/56492.

### CODE REVIEW

File: v21.1/known-limitations.md [hunk 1]
```diff
@@ -386,6 +443,10 @@ When inserting/updating all columns of a table, and the table has no secondary i
 
 This issue is particularly relevant when using a simple SQL table of two columns to [simulate direct KV access](sql-faqs.html#can-i-use-cockroachdb-as-a-key-value-store). In this case, be sure to use the `UPSERT` statement.
 
+### Size limits on statement input from SQL clients
```
Comment (ericharmeling) [inline v21.1/known-limitations.md:446]: @bdarnell

File: v21.1/known-limitations.md [hunk 2]
```diff
@@ -130,34 +285,11 @@ CockroachDB supports efficiently storing and querying [spatial data](spatial-dat
 
     [Tracking GitHub Issue](https://github.com/cockroachdb/cockroach/issues/55227)
 
-### Collation names that include upper-case or hyphens may cause errors
+- CockroachDB does not support using [schema name prefixes](sql-name-resolution.html#how-name-resolution-works) to refer to [data types](data-types.html) with type modifiers (e.g., `public.geometry(linestring, 4326)`). Instead, use fully-unqualified names to refer to data types with type modifiers (e.g., `geometry(linestring,4326)`).
```
Comment (ericharmeling) [inline v21.1/known-limitations.md:288]: @otan

File: v21.1/known-limitations.md [hunk 3]
```diff
@@ -6,6 +6,159 @@ toc: true
 
 This page describes newly identified limitations in the CockroachDB {{page.release_info.version}} release as well as unresolved limitations identified in earlier releases.
 
+## New limitations
+
+### CockroachDB does not properly optimize some left and inverted joins with inverted indexes
+
```
Comment (ericharmeling) [inline v21.1/known-limitations.md:12]: @rytaft

File: v21.1/bytes.md [hunk 4]
```diff
@@ -30,6 +30,10 @@ is otherwise expected.
 
 The size of a `BYTES` value is variable, but it's recommended to keep values under 1 MB to ensure performance. Above that threshold, [write amplification](https://en.wikipedia.org/wiki/Write_amplification) and other considerations may cause significant performance degradation.  
 
+{{site.data.alerts.callout_success}}
+If your application requires large binary input in single queries, you can store the blobs somewhere your client can access them (using a cloud storage service, for example), and then reference their addresses from a statement.
```
Comment (ericharmeling) [inline v21.1/bytes.md:34]: @bdarnell

**Comment by ericharmeling** [comment]
Adding relevant engineers to review specific limitations. Thank you!

File: v21.1/known-limitations.md (hunk 3 above)
Comment (rytaft) [reply v21.1/known-limitations.md:12]: just a couple of nits, otherwise LGTM - thanks!

File: v21.1/known-limitations.md [hunk 5]
```diff
@@ -6,6 +6,159 @@ toc: true
...
+[Left joins](joins.html#left-outer-joins) and [inverted joins](joins.html#inverted-joins) involving [`JSONB`](jsonb.html), [`ARRAY`](array.html), or [spatial-typed](spatial-data.html) columns with a multi-column or [partitioned](partition-by.html) [inverted index](inverted-indexes.html) will not take advantage of the index if the prefix columns of the index are unconstrained, or if they are constrained to multiple, constant values.
```
Comment (rytaft) [inline v21.1/known-limitations.md:13]: nit: this should say left joins and anti joins (it's true the relevant algorithm is inverted join, but the limitation is only for left and anti joins, not inner or semi joins).

File: v21.1/known-limitations.md [hunk 6]
```diff
@@ -6,6 +6,159 @@ toc: true
...
+
+To work around this limitation, make sure that the prefix columns of the index are either constrained to single constant values, or are the same as the input columns.
```
Comment (rytaft) [inline v21.1/known-limitations.md:15]: nit: "the same as input columns" is misleading -- I think this should be referring to equality conditions. If the prefix columns are not held constant, they should be part of equality conditions (e.g., col1 = col2), where col1 is a prefix column and col2 is an input column.

**Review by jseldess — APPROVED** [review]
:lgtm:, after addressing feedback form others. 

@ericharmeling, please make sure that any corresponding `cockroach` GH issues have the `docs-done` label added. That gives us a quick way to know that a known limitations has been documented. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10568#-:-M_qT9zK5eepUzOIVeAt:b-pnb0ov)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bdarnell, @ericharmeling, and @otan)

___
*[v20.1/known-limitations.md, line 485 at r2](https://reviewable.io/reviews/cockroachdb/docs/10568#-M_qRtAh-zeqnZ0bBE4W:-M_qRtAh-zeqnZ0bBE4X:brki81x) ([raw file](https://github.com/cockroachdb/docs/blob/ac5c59be5fa4da8755ce5a763865f46eab579956/v20.1/known-limitations.md#L485)):*
> ```Markdown
> This issue is particularly relevant when using a simple SQL table of two columns to [simulate direct KV access](sql-faqs.html#can-i-use-cockroachdb-as-a-key-value-store). In this case, be sure to use the `UPSERT` statement.
> 
> ### Size limits on statement input from SQL clients
> ```

Do we have a GH issue to link to here, or is this something we don't plan to fix/change over time?

Also, add final period.
___
*[v20.2/known-limitations.md, line 352 at r2](https://reviewable.io/reviews/cockroachdb/docs/10568#-M_qS8Ij52MnWke0DbMC:-M_qS8Ij52MnWke0DbMD:b-y60rnc) ([raw file](https://github.com/cockroachdb/docs/blob/ac5c59be5fa4da8755ce5a763865f46eab579956/v20.2/known-limitations.md#L352)):*
> ```Markdown
> This issue is particularly relevant when using a simple SQL table of two columns to [simulate direct KV access](sql-faqs.html#can-i-use-cockroachdb-as-a-key-value-store). In this case, be sure to use the `UPSERT` statement.
> 
> ### Size limits on statement input from SQL clients
> ```

Same as before: Do we have a GH issue to link to here, or is this something we don't plan to fix/change over time?.

Also, add final period. 
___
*[v21.1/bytes.md, line 34 at r2](https://reviewable.io/reviews/cockroachdb/docs/10568#-M_btRzRh2_u2VYUfTQH-r2-34:-M_qT-N42BrB-fiE2ned:b2nreh6) ([raw file](https://github.com/cockroachdb/docs/blob/ac5c59be5fa4da8755ce5a763865f46eab579956/v21.1/bytes.md#L34)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

@bdarnell 
</blockquote></details>

Same as above.

**Review by otan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10568)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bdarnell and @ericharmeling)

___
*[v21.1/known-limitations.md, line 288 at r2](https://reviewable.io/reviews/cockroachdb/docs/10568#-M_btRzRh2_u2VYUfTQJ-r2-288:-M_rPJ9REzHAfir5MDVP:b-li568r) ([raw file](https://github.com/cockroachdb/docs/blob/ac5c59be5fa4da8755ce5a763865f46eab579956/v21.1/known-limitations.md#L288)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

@otan 
</blockquote></details>

This should be the case for 20.2 as well.

**Review by ericharmeling — COMMENTED** [review]
TFTR @jseldess, @rytaft, @otan!

I believe I've addressed everyone's comments.

@jseldess I added those `docs-done` labels, now that this is approved.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10568#-:-M_ucNFwBUCb4CARpvA4:bxvijyv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @bdarnell and @jseldess)

___
*[v20.1/known-limitations.md, line 485 at r2](https://reviewable.io/reviews/cockroachdb/docs/10568#-M_qRtAh-zeqnZ0bBE4W:-M_uZf_K6WKIIBfsUNi5:b-99yv9z) ([raw file](https://github.com/cockroachdb/docs/blob/ac5c59be5fa4da8755ce5a763865f46eab579956/v20.1/known-limitations.md#L485)):*
> Do we have a GH issue to link to here, or is this something we don't plan to fix/change over time?

I don't believe that there is a GitHub issue to resolve this specific limitation.

A quote from [this thread](https://cockroachlabs.slack.com/archives/CHVV403F0/p1590414556235900?thread_ts=1590406019.235800&cid=CHVV403F0):

"... the 16 MiB limit is in our implementation of the postgres network protocol. it's a magic number we could change, but we've mainly discussed that in the context of lowering it, not raising it. "

So it appears that any changes made would lower that hard limit (making the limitation more restrictive), but not remove it.

> Also, add final period.

Done.
___
*[v20.2/known-limitations.md, line 352 at r2](https://reviewable.io/reviews/cockroachdb/docs/10568#-M_qS8Ij52MnWke0DbMC:-M_u_0_B9WGAavQDFy6U:bne4p88) ([raw file](https://github.com/cockroachdb/docs/blob/ac5c59be5fa4da8755ce5a763865f46eab579956/v20.2/known-limitations.md#L352)):*
> Do we have a GH issue to link to here, or is this something we don't plan to fix/change over time?

See my comment above.

> Also, add final period.

Done.
___
*[v21.1/bytes.md, line 34 at r2](https://reviewable.io/reviews/cockroachdb/docs/10568#-M_btRzRh2_u2VYUfTQH-r2-34:-M_ubzmn6KaHGvHylUdG:bnh5vpt) ([raw file](https://github.com/cockroachdb/docs/blob/ac5c59be5fa4da8755ce5a763865f46eab579956/v21.1/bytes.md#L34)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Same as above.
</blockquote></details>

This is a tip from [that same thread about size limits on SQL statement input](https://cockroachlabs.slack.com/archives/CHVV403F0/p1590423578236500?thread_ts=1590406019.235800&cid=CHVV403F0). 

I discussed with Ben [here](https://github.com/cockroachdb/docs/issues/7524#issuecomment-840654004), and I agree with his suggestion to make this tip an extension of the size recommendations on this particular type, rather than overcomplicate the stated limitation on SQL input size.
___
*[v21.1/known-limitations.md, line 13 at r2](https://reviewable.io/reviews/cockroachdb/docs/10568#-M_btRzRh2_u2VYUfTQJ-r2-13:-M_ufGCB2X8kNvQrBdX9:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/ac5c59be5fa4da8755ce5a763865f46eab579956/v21.1/known-limitations.md#L13)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: this should say left joins and anti joins (it's true the relevant algorithm is inverted join, but the limitation is only for left and anti joins, not inner or semi joins).
</blockquote></details>

Done.
___
*[v21.1/known-limitations.md, line 15 at r2](https://reviewable.io/reviews/cockroachdb/docs/10568#-M_btRzRh2_u2VYUfTQJ-r2-15:-M_ujp133G8iyakAfgy3:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/ac5c59be5fa4da8755ce5a763865f46eab579956/v21.1/known-limitations.md#L15)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: "the same as input columns" is misleading -- I think this should be referring to equality conditions. If the prefix columns are not held constant, they should be part of equality conditions (e.g., col1 = col2), where col1 is a prefix column and col2 is an input column.
</blockquote></details>

Done.
___
*[v21.1/known-limitations.md, line 288 at r2](https://reviewable.io/reviews/cockroachdb/docs/10568#-M_btRzRh2_u2VYUfTQJ-r2-288:-M_ucE1o-LKq1BpFMGnf:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/ac5c59be5fa4da8755ce5a763865f46eab579956/v21.1/known-limitations.md#L288)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

This should be the case for 20.2 as well.
</blockquote></details>

Done.

**Review by rytaft — COMMENTED** [review]
> Reviewed 1 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10568)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @bdarnell, @ericharmeling, and @jseldess)

___
*[v21.1/known-limitations.md, line 613 at r3](https://reviewable.io/reviews/cockroachdb/docs/10568#-M_unrYS9bae-NO4YNKw:-M_unrYTD4lFmSa3zVjc:bprfzby) ([raw file](https://github.com/cockroachdb/docs/blob/ad2b4219d04d21528dafec29d52faec4646a9362/v21.1/known-limitations.md#L613)):*
> ```Markdown
> ### Inverted indexes cannot be partitioned
> 
> CockroachDB does not support partitioning inverted indexes, including [spatial indexes](spatial-indexes.html).
> ```

This is no longer a limitation. 21.1 supports partitioning of inverted indexes.

**Review by ericharmeling — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10568)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @bdarnell, @jseldess, and @rytaft)

___
*[v21.1/known-limitations.md, line 613 at r3](https://reviewable.io/reviews/cockroachdb/docs/10568#-M_unrYS9bae-NO4YNKw:-M_utM0t6bXB8pabtyYd:b69upvu) ([raw file](https://github.com/cockroachdb/docs/blob/ad2b4219d04d21528dafec29d52faec4646a9362/v21.1/known-limitations.md#L613)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This is no longer a limitation. 21.1 supports partitioning of inverted indexes.
</blockquote></details>

Good catch! 

Removed.

**Comment by ericharmeling** [comment]
Merging so this can go out with the GA 21.1 release notes - @bdarnell, feel free to leave feedback after the merge if you'd like to. I'm happy to make updates as-needed.

**Pull request merged**